crates/fuel-core/src/state/historical_rocksdb.rs

```diff
@@ -300,12 +354,96 @@ where
 
         Ok(())
     }
+
+    /// Migrates a ModificationHistory key-value pair from V1 to V2.
+    /// The migration fails if a concurrent transaction which rollbacks
+    /// to the `height` being migrated commits before this
+    /// migration transaction commits.
+    pub fn migrate_modifications_history_at_height(
+        &self,
+        height: u64,
+    ) -> StorageResult<()> {
+        let mut migration_transaction = StorageTransaction::transaction(
+            &self.db,
+            ConflictPolicy::Fail,
```

Comment:
Conflict policy works only when you try to commit one in-memory transaction into another. It doesn't work with the Rocksdb.

The migration process should accumulate changes in the memory and dump them inside of the `commit_changes` function(when upper services like block importer or relayer call this function).

Reply:
Thanks. In the latest version of the MR I switched to having cumulative changes in the rocksDB state, and I handle race conditions by locking on them for the whole duration of the migration/rollback/commit_changes function. Unfortunately I cannot use more fine-grained guards because of potential consistency issues (explanation in the comments).